### PR TITLE
[MDEP-799] tree: add optional output type json

### PR DESCRIPTION
**Pull request by MartinWitt**
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MDEP) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.

I will squash and update the commit in the end. There is still open discussion. 

 - [X] Format the pull request title like `[MDEP-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MDEP-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

 - [X] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

I sent the mail for this a few minutes ago so it should be signed soon.

I tried to understand the way you write unit/integration tests but it looks a bit complicated. What is the correct way to add a testcase? Best case, I can add a pom and check if the model written as json -> read by any JSON parser is equal to the model before.

The JSON files currently exist for you to see the current printed format. It will be deleted afterward.

This PR adds JSON output type to the maven-depdenceny-plugin with the goal `tree`.

Supersedes #207

### CODE REVIEW

**Comment by MartinWitt** [comment]
CLA is now signed and accepted.

File: src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java [hunk 1]
```diff
@@ -380,6 +380,8 @@ public DependencyNodeVisitor getSerializingDependencyNodeVisitor(Writer writer)
             return new TGFDependencyNodeVisitor(writer);
         } else if ("dot".equals(outputType)) {
             return new DOTDependencyNodeVisitor(writer);
+        } else if ("json".equals(outputType)) {
+            return new JsonDependencyNodeVisitor(writer);
         } else {
             return new SerializingDependencyNodeVisitor(writer, toGraphTokens(tokens));
```
Comment (MartinWitt) [inline src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java:386]: In the other PR someone asked for a switch case instead. Shall I convert this to a switch statement?

**Comment by MartinWitt** [comment]
As I am a first-time contributor, someone has to approve the workflow run before we see CI results.

**Comment by MartinWitt** [comment]
@elharo could you approve the worflow run?

File: src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    /**
+     * Creates a new instance of {@link JsonDependencyNodeVisitor}. The writer will be used to write the output.
+     * @param writer  the writer to write to
+     */
+    public JsonDependencyNodeVisitor(Writer writer) {
+        super(writer);
+    }
+
+    @Override
+    public boolean visit(DependencyNode node) {
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:46]: Is it possible for this to et stuck in an infinite recursion with a maliciously hand-crafted dependency node? Or for that matter, with a non-maliciious but buggy circular dependency tree (which does happen)?

File: src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.shared.dependency.graph.DependencyNode;
+import org.apache.maven.shared.dependency.graph.traversal.DependencyNodeVisitor;
+
+/**
+ * A dependency node visitor that serializes visited nodes to a writer using the JSON format.
+ */
+public class JsonDependencyNodeVisitor extends AbstractSerializingVisitor implements DependencyNodeVisitor {
+
+    private static final String LINE_SEPARATOR = "\n";
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:34]: inline this, a simple "\n" is clearer

File: src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        sb.append("}").append(LINE_SEPARATOR);
+        try {
+            writer.write(sb.toString());
+        } catch (IOException e) {
+            throw new RuntimeException("Error while writing json output", e);
+            // TODO: handle exception maybe throw runtime exception or mojo exception?
+        }
+    }
+    /**
+     * Appends the node and it's children to the string builder.
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:72]: its

File: src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        }
+    }
+    /**
+     * Appends a key value pair to the string builder.
+     * @param sb  the string builder to append to
+     * @param indent  the current indent level
+     * @param key  the key used as json key
+     * @param value  the value used as json value
+     */
+    private void appendKey(StringBuilder sb, int indent, String key, String value) {
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:139]: appendKeyValue

File: src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java (hunk 1 above)
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java:386]: no, this is fine

File: src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java [hunk 6]
```diff
@@ -127,6 +127,24 @@ public void _testTreeTGFSerializing() throws Exception {
         assertTrue(findString(contents, "testGroupId:release:jar:1.0:compile"));
     }
 
+    /**
+     * Test the JSON format serialization
+     */
+    public void _testTreeJsonSerialzing() throws Exception {
+        List<String> contents = runTreeMojo("tree1.json", "json");
+        assertTrue(findString(contents, "\"testGroupId\": \"project\""));
```
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java:135]: not sure why findString is used here. contains is simple and more obvious

Comment (MartinWitt) [reply src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java:135]: I tried to keep the test case in the same style as the rest of the class. But I can change it.

**Comment by pombredanne** [comment]
@MartinWitt what's left to do to get this through?
FWIW, there is a cottage industry of smalls tools and scripts that are parsing this plugin output .... it would be awesome to have a structure JSON output! 
Anything to help you need move this forward?

File: src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.shared.dependency.graph.DependencyNode;
+import org.apache.maven.shared.dependency.graph.traversal.DependencyNodeVisitor;
+
+/**
+ * A dependency node visitor that serializes visited nodes to a writer using the JSON format.
+ */
+public class JsonDependencyNodeVisitor extends AbstractSerializingVisitor implements DependencyNodeVisitor {
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:32]: Does this need to be public? It's easier to evolve and iterate on if it's not.

File: src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    private void writeRootNode(DependencyNode node, Writer writer) {
+        int indent = 2;
+        StringBuilder sb = new StringBuilder();
+        sb.append("{").append("\n");
+        writeNode(indent, node, sb);
+        sb.append("}").append("\n");
+        try {
+            writer.write(sb.toString());
+        } catch (IOException e) {
+            throw new RuntimeException("Error while writing json output", e);
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:66]: needs a more specific exception. Looking at this now I notice that the superclass and interface are badly designed. They use a PrintWriter which swallows exceptions. That might need to be fixed.

Comment (pombredanne) [reply src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:66]: What about just not catch an exception here and leaving this business to the callers.... fixing the PrintWriter could then be done in another patch and not here.

**Comment by pombredanne** [comment]
Looking at the details of this PR, it feels to me that crafting the JSON by hand as done here feels like problems in the making with encoding or else. Should not this use a proper JSON library of sorts? Is not there one bundled in the standard Maven or Java and this plugin may therefore always have have one on hand? 

(Sorry for these likely dumb questions: my Java skills need to brushing off)

**Comment by yahavi** [comment]
If it's useful to anyone, we've developed a Maven dependency tree plugin that generates a JSON dependency tree. Check it out at: https://github.com/jfrog/maven-dep-tree.

**Comment by monperrus** [comment]
@pombredanne @elharo if we are able to drive to merge, we can put more effort on this PR.

File: src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java (hunk 8 above)
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:66]: Fixing PrintWriter should be a different PR but that might be a prerequisite for this PR. I'm not sure. It seems fine to let this method throw IOException and handle it further up the stack. RuntimeException is not OK.

File: src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java (hunk 6 above)
Comment (elharo) [reply src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java:135]: contains is clearer

File: src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java (hunk 2 above)
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/tree/JsonDependencyNodeVisitor.java:45]: add a test for this case, see what works

**Review by elharo — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java (hunk 6 above)
Comment (LogFlames) [reply src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java:135]: I want to double check: do you mean `contents.contains("...")`? 
Due to indentation the string might not be in the list of strings, but a substring of an element, which the `findString`-method seems to be created for in the same file.

**Comment by LogFlames** [comment]
Hi, 
I'm working on getting this merged and incorporating the requested changes. Unfortunately I don't have access to Martin's fork and will need to create a new PR. Will build on Martin's branch and try to update all changes beforehand to make the transition as smooth as possible.

File: src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java (hunk 6 above)
Comment (elharo) [reply src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java:135]: That makes sense.

**Comment by jira-importer** [comment]
Resolve #1310

**Pull request closed without merging**